### PR TITLE
ref: object patcher and patch collector

### PR DESCRIPTION
**Pull request by diafour**
#### Overview

- add ignoreIfMissing option for patch operations
- refactor object patcher and patch collector for simpler use In Go hooks

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/kube/object_patch/operation.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		//mergePatch, err := json.Marshal(spec.MergePatch)
+		//if err != nil {
+		//	applyErrors = multierror.Append(applyErrors, err)
+		//	continue
+		//}
+		//operationError = o.Patch(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+		//	WithSubresource(spec.Subresource),
+		//	UseMergePatch(mergePatch),
+		//	WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		//)
```
Comment (diafour) [inline pkg/kube/object_patch/operation.go:183]: PatchCollector was able to return an error if the patch is not valid. Now, PatchCollector saves the patch as-is. Patch content is handled by the ObjectPatcher during ExecuteOperations after hook execution.

File: pkg/kube/object_patch/operation.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		namespace:  namespace,
+		name:       name,
+		patch:      jsonPatch,
+		patchType:  types.JSONPatchType,
+	}
+	op.applyOptions(options...)
+	return op
+}
+
+func NewFilterPatchOperation(filter func(*unstructured.Unstructured) (*unstructured.Unstructured, error), apiVersion, kind, namespace, name string, options ...OperationOption) *Operation {
```
Comment (diafour) [inline pkg/kube/object_patch/operation.go:243]: There is no NewJQPatchOperation. JQPatch is now available for shell hooks and not available for Go hooks.

File: pkg/kube/object_patch/options.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+type OperationOption func(operation *Operation)
+
+// WithSubresource options specifies a subresource to operate on.
+func WithSubresource(subresource string) OperationOption {
+	return func(operation *Operation) {
+		operation.subresource = subresource
+	}
+}
+
+func UseFilterFunc(filterFunc func(*unstructured.Unstructured) (*unstructured.Unstructured, error)) OperationOption {
```
Comment (diafour) [inline pkg/kube/object_patch/options.go:18]: Oops. I should get rid of these methods.

File: pkg/kube/object_patch/patch_collector.go [hunk 4]
```diff
@@ -1,156 +1,48 @@
...
-		Name:        name,
-		Subresource: subresource,
-		FilterFunc:  filterFunc,
-	}
-	dop.patchOperations = append(dop.patchOperations, op)
-
-	return nil
+	apiVersion, kind, namespace, name string, options ...OperationOption,
+) {
+	dop.patchOperations = append(dop.patchOperations, NewFilterPatchOperation(filterFunc, apiVersion, kind, namespace, name, options...))
```
Comment (diafour) [inline pkg/kube/object_patch/patch_collector.go:38]: should use `dop.add`

**Review by yalosev — CHANGES_REQUESTED** [review]
Don't you want to separate CreateOptions/DeleteOptions/UpdateOptions. It will make func signatures more obvious

**Comment by diafour** [comment]
> Don't you want to separate CreateOptions/DeleteOptions/UpdateOptions. It will make func signatures more obvious

Done!

**Review by yalosev — APPROVED** [review]
LGTM

**Comment by diafour** [comment]
Approved in deckhouse.

**Pull request merged**